scripts/lib/CIME/XML/component.py

```diff
@@ -116,14 +118,74 @@ def _get_value_match(self, node, attributes=None, exact_match=False):
         return match_value
 
     #pylint: disable=arguments-differ
-    def get_description(self, compsetname):
+    def get_description(self, compsetname, comp_class=None):
+        comp_class = comp_class.lower()
         rootnode = self.get_node("description")
-        desc_nodes = self.get_nodes("desc", root=rootnode)
         desc = ""
-        for node in desc_nodes:
-            compsetmatch = node.get("compset")
-            if re.search(compsetmatch, compsetname):
-                desc += node.text
+        desc_nodes = self.get_nodes("desc", root=rootnode)
+        if self.get_version() == 3.0:
+            expect(comp_class is not None,"comp_class argument required for version3 files")
+            modifier_mode = rootnode.get('modifier_mode')
+            if modifier_mode is None:
+                modifier_mode = '*'
+                expect(modifier_mode in ('*','1','?','+'),
+                       "Invalid modifier_mode {} in file {}".format(modifier_mode, self.filename))
+
+            optiondesc = {}
+            if comp_class == "forcing":
+                for node in desc_nodes:
+                    forcing = node.get('forcing')
+                    if forcing is not None and compsetname.startswith(forcing):
+                        desc = node.text
+                        return desc
+
+            # first pass just make a hash of the option descriptions
+            for node in desc_nodes:
+                option = node.get('option')
+                if option is not None:
+                    optiondesc[option] = node.text
+            #second pass find a comp_class match
```

Comment:
I couldn't understand what the following code is doing, at a high level (so I haven't given it much of a review). This could use some high-level documentation and/or examples. Ideally, there would also be some unit tests - either for this whole get_description method or for the core algorithm (below), which could be extracted into its own method.